pkg/operation/botanist/component/vpnshoot/vpnshoot.go

```diff
@@ -0,0 +1,515 @@
+				},
+			},
+		}
+	)
+	utilruntime.Must(references.InjectAnnotations(deployment))
+
+	if !v.values.ReversedVPNEnabled {
+		service = &corev1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "vpn-shoot",
+				Namespace: metav1.NamespaceSystem,
+				Labels:    map[string]string{v1beta1constants.LabelApp: LabelValue},
+			},
+			Spec: corev1.ServiceSpec{
+				Selector: map[string]string{v1beta1constants.LabelApp: LabelValue},
+				Type:     corev1.ServiceTypeLoadBalancer,
+				Ports: []corev1.ServicePort{
+					{
+						Name:       "openvpn",
+						Port:       servicePort,
+						TargetPort: intstr.FromInt(int(containerPort)),
+						Protocol:   corev1.ProtocolTCP,
+					},
+				},
+			},
+		}
+		clusterRole = &rbacv1.ClusterRole{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "system:gardener.cloud:vpn-seed",
+			},
+			Rules: []rbacv1.PolicyRule{
+				{
+					APIGroups:     []string{""},
+					Resources:     []string{"services"},
+					ResourceNames: []string{"vpn-shoot"},
+					Verbs:         []string{"get"},
+				},
+			},
+		}
+
+		clusterRoleBinding = &rbacv1.ClusterRoleBinding{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:        "system:gardener.cloud:vpn-seed",
+				Annotations: map[string]string{resourcesv1alpha1.DeleteOnInvalidUpdate: "true"},
+			},
+			RoleRef: rbacv1.RoleRef{
+				APIGroup: rbacv1.GroupName,
+				Kind:     "ClusterRole",
+				Name:     clusterRole.Name,
+			},
+			Subjects: []rbacv1.Subject{
+				{
+					Kind: rbacv1.UserKind,
+					Name: "vpn-seed",
+				},
+			},
+		}
+	}
+
+	if v.values.VPAEnabled {
+		vpaUpdateMode := autoscalingv1beta2.UpdateModeAuto
+		vpa = &autoscalingv1beta2.VerticalPodAutoscaler{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "vpn-shoot",
+				Namespace: metav1.NamespaceSystem,
+			},
+			Spec: autoscalingv1beta2.VerticalPodAutoscalerSpec{
+				TargetRef: &autoscalingv1.CrossVersionObjectReference{
+					APIVersion: appsv1.SchemeGroupVersion.String(),
+					Kind:       "Deployment",
+					Name:       deploymentName,
+				},
+				UpdatePolicy: &autoscalingv1beta2.PodUpdatePolicy{
+					UpdateMode: &vpaUpdateMode,
+				},
+				ResourcePolicy: &autoscalingv1beta2.PodResourcePolicy{
+					ContainerPolicies: []autoscalingv1beta2.ContainerResourcePolicy{
+						{
+							ContainerName: autoscalingv1beta2.DefaultContainerResourcePolicy,
+							MinAllowed: corev1.ResourceList{
+								corev1.ResourceCPU:    resource.MustParse("10m"),
+								corev1.ResourceMemory: resource.MustParse("10Mi"),
+							},
+						},
+					},
+				},
+			},
+		}
+	}
+
+	return registry.AddAllAndSerialize(
+		secret,
+		secretTLSAuth,
+		secretDH,
+		serviceAccount,
+		networkPolicy,
+		deployment,
+		clusterRole,
+		clusterRoleBinding,
+		service,
+		vpa,
+	)
+}
+
+// Secrets is collection of secrets for the vpn-shoot.
+type Secrets struct {
+	// TLSAuth is a secret containing the tls auth credentials.
+	TLSAuth component.Secret
+	// DH is a secret containing the dh credentials.
+	DH component.Secret
```

Comment:
This secret is only needed when ReversedVPN is disabled, is it? Then we probably should make it 
```suggestion
	DH *component.Secret
```